### PR TITLE
[HLO] Fix nested dynamic-slice and dynamic-update-slice fold index clamping (miscompile)

### PR DESCRIPTION
**Pull request by EylonKrause**
#ai-generated

## Problem

Two sibling folds in `AlgebraicSimplifier` compose the start indices of nested dynamic-slice / dynamic-update-slice ops incorrectly, so an **out-of-range but valid** index (HLO `DynamicSlice`/`DynamicUpdateSlice` clamp their start indices per the spec) is placed at the wrong offset — a silent miscompile. For in-range indices every clamp is a no-op, which is why it went unnoticed.

### 1. `HandleDynamicSlice` — `ds(ds(x, inner_id){inner}, id){outer}`

Correct composed start into `x` is `clamp(inner_id, 0, dim - inner_size) + clamp(id, 0, inner_size - outer_size)`. The fold (a) clamped the **inner** index with the **outer** slice size (`dim - outer_size` instead of `dim - inner_size`), and (b) added the **outer** index `id` **unclamped**.

Example `x=f32[10]`, `inner=ds(x,inner_id){4}`, `outer=ds(inner,id){2}`, `inner_id=8, id=0`: correct `{x6,x7}`, folded `{x8,x9}`.

### 2. `HandleDynamicUpdateSlice` — `dus(a, dus(ds(a, id){ds}, c, inner_id), id)`

Correct composed update start is `clamp(id, 0, dim - ds_size) + clamp(inner_id, 0, ds_size - c_size)`. The inner index is clamped correctly here, but the **outer** index `id` is added **unclamped**; the retained DUS only clamps the sum to `[0, dim - c_size]`, weaker than the `[0, dim - ds_size]` the original enforces (via both the inner `ds` and outer `dus`).

Example `a=f32[10]`, `ds{4}`, `c=f32[2]`, `id=100, inner_id=0`: correct writes `c` at `[6,8)`, folded writes at `[8,10)`.

## Fix

Clamp each index with its own bound before adding, in both folds. `inner_size >= outer_size` / `ds_size >= c_size` is guaranteed (the outer op slices/updates within the inner tensor), so the added clamp bound is non-negative, and the composed start lies in the range the retained op clamps to (making that clamp a no-op and the fold exact).

## Testing

XLA doesn't build on my dev box, so I verified the **index arithmetic** against ground-truth nested-slice semantics with standalone programs:

- dynamic-slice: 2,760 `(inner_id, id)` combinations × 7 `(dim, inner, outer)` configs — the fix matches truth for **all** (0 mismatches); the current code is wrong in **1,284**.
- dynamic-update-slice: 3,465 combinations × 7 configs — fix **0** mismatches, current **1,204**.

Added two structural regression tests (`DynamicSliceOfDynamicSliceClampsBothIndices`, `DynamicUpdateSliceOfDynamicUpdateSliceClampsBothIndices`) asserting the folded op clamps **both** start indices (`Add(Clamp(..), Clamp(..))`), which fail on the current code (the outer index is added unclamped). I couldn't run the C++ suite here — happy to adjust the matchers if CI flags them.

Disclosure: this contribution was authored with an AI coding assistant (Claude) and reviewed before submission.

### CODE REVIEW

**Comment by golechwierowicz** [comment]
Do you have a JAX reproducer? The issue sounds very theoretical to me, and in general we do not accept such. Please check out our [contribution guidelines](https://github.com/openxla/xla/blob/main/docs/contributing.md#ai-generated-contributions)

**Comment by EylonKrause** [comment]
Fair enough — here's a JAX repro showing it's a real miscompile, not just a structural nit. Same function, `jit` vs eager diverge on valid inputs:

```python
import jax, jax.numpy as jnp, numpy as np
from jax import lax

def f(x, i, j):
    inner = lax.dynamic_slice(x, (i,), (4,))
    return lax.dynamic_slice(inner, (j,), (2,))

x = jnp.arange(10, dtype=jnp.float32)
for i, j in [(8, 0), (0, 100), (9, 0)]:
    print((i, j), np.asarray(f(x, i, j)), np.asarray(jax.jit(f)(x, i, j)))
```

Output (jax/jaxlib 0.11.1, CPU):

```
(8, 0)   [6. 7.]   [8. 9.]
(0, 100) [2. 3.]   [8. 9.]
(9, 0)   [6. 7.]   [8. 9.]
```

Eager is right, `jit` is wrong. It only bites when a start index is out of range — but those are defined inputs: `dynamic_slice` clamps the start to `[0, dim - slice_size]` (that's why eager gives the right answer). The nested-slice fold in AlgebraicSimplifier composes the two starts without clamping each to its own range, so once the simplifier fires the result changes. Data-dependent indices that can run past the end are a normal pattern (sliding windows, `x[i:i+k]` with `i` from the program), so it isn't a hand-crafted corner.

The fix is a no-op for in-range indices — every clamp I add collapses away when the index is already valid — so it can't regress the common case. I swept every index combination against the true nested-slice semantics too: the patched fold matches for all of them, the current one is wrong in ~46%.

I've added the `#ai-generated` tag to the description. It was AI-assisted, but I went through the fold and verified it myself — the repro above and the sweep are how I convinced myself before opening it.

**Comment by derdrdirk** [comment]
Closing due to [contribution guidelines](https://github.com/openxla/xla/blob/main/docs/contributing.md#ai-generated-contributions)
